docs: add best practices documentation (#779)
Azure/aad-pod-identity: docs/readmes/README.best-practices.md

# Best Practices

This document highlights the best practices when using aad-pod-identity.

## Retry on token retrieval

> aad-pod-identity retrieves AAD tokens on behalf of your workload by intercepting token requests to the Instance Metadata Service (IMDS) endpoint (169.254.169.254). However, it does not perform any sort of token caching and retries. We believe that either the users or the SDK that your application is using is responsible for token management.

Reviewer: nit:s/AAD tokens/access tokens

> Note that there is a brief window between when your application started running and the identity gets assigned to the underlying node. If your application tried to retrieve a token during that period, it will fail. The following table describes the duration that you could expect based on your cluster's node type:

Reviewer: We should also add a note about the in built NMI retries we have in place. If the token request times out before these retries succeed, then token it'll fail. Also, add a note that this is only the first time the identity is being assigned to the underlying node. Subsequent token requests will not incur this delay.

| Node Type                 | Identity Assignment Duration |
|---------------------------|------------------------------|
| Virtual Machine           | 10 - 20 seconds              |
| Virtual Machine Scale Set | 40 - 60 seconds              |

> Without retry, your application will not able to retrieve AAD tokens because the identity assignment operation is still in progress. Usually, retry logic is already implemented internally in most of Azure's SDKs to prevent your application from erroring out. For example, [azure-sdk-for-go](https://github.com/Azure/azure-sdk-for-go) by default performs five retries with exponential backoff when fetching AAD token from IMDS.

Reviewer: nit: s/AAD/access

> Alternatively, you could set up a simple init container to probe for successful identity assignment by MIC and prevent applications from starting immaturely. Here is an example init-container with Azure CLI.

Reviewer: s/immaturely/prematurely

```yaml
...
initContainers:
- name: init
  image: mcr.microsoft.com/azure-cli
  command:
    - sh
    - -c
    - az login --identity --debug
...
```

## A pod using multiple AzureIdentities

In some scenarios, you might want to assign multiple `AzureIdentities` to a workload. Here is an example of how to achieve that:

### AzureIdentities

```yaml
apiVersion: aadpodidentity.k8s.io/v1
kind: AzureIdentity
metadata:
  name: az-id-1
spec:
  type: 0
  resourceID: <ResourceID of az-id-1>
  clientID: <ClientID of az-id-1>
```

```yaml
apiVersion: aadpodidentity.k8s.io/v1
kind: AzureIdentity
metadata:
  name: az-id-2
spec:
  type: 0
  resourceID: <ResourceID of az-id-2>
  clientID: <ClientID of az-id-2>
```

### AzureIdentityBinding

```yaml
apiVersion: aadpodidentity.k8s.io/v1
kind: AzureIdentityBinding
metadata:
  name: az-id-1-binding
spec:

>   azureIdentity: az-id-1-binding

Reviewer: ```suggestion
  azureIdentity: az-id-1
```

  selector: az-id-combined
```

```yaml
apiVersion: aadpodidentity.k8s.io/v1
kind: AzureIdentityBinding
metadata:
  name: az-id-2-binding
spec:

>   azureIdentity: az-id-2-binding

Reviewer: ```suggestion
  azureIdentity: az-id-2
```

  selector: az-id-combined
```

### Pod

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: demo
  labels:
    aadpodidbinding: az-id-combined
...
```

## Pods using unauthorized AzureIdentities

By default, aad-pod-identity matches pods with `AzureIdentities` across all namespaces. That means that a malicious pod could assign itself an unauthorized pod identity label and acquire a token with that particular `AzureIdentity`. This scenario can be mitigated by performing the following:

> - Deploy aad-pod-identity with [namespaced mode](README.namespaced.md). This will restrict pods from binding with `AzureIdentities` across different namespaces.

Reviewer: Lets add a note that in namespaced mode, the pod, AzureIdentity and AzureIdentityBinding all need to be in the same namespace

- Set up Kubernetes-native RBAC to restrict access and creation of `AzureIdentities` across sensitive namespaces.

- Set up RBAC on Azure to restrict access and creation of user-assigned identities.

> ## Using aad-pod-identity in various workloads

Reviewer: ```suggestion
## Using aad-pod-identity in different Kubernetes workloads
```

```yaml
apiVersion: aadpodidentity.k8s.io/v1
kind: AzureIdentity
metadata:
  name: az-id-1
spec:
  type: 0
  resourceID: <ResourceID of az-id-1>
  clientID: <ClientID of az-id-1>
```

```yaml
apiVersion: aadpodidentity.k8s.io/v1
kind: AzureIdentity
metadata:
  name: az-id-2
spec:
  type: 0
  resourceID: <ResourceID of az-id-2>
  clientID: <ClientID of az-id-2>
```

### AzureIdentityBinding

```yaml
apiVersion: aadpodidentity.k8s.io/v1
kind: AzureIdentityBinding
metadata:
  name: az-id-1-binding
spec:
  azureIdentity: az-id-1
  selector: az-id-combined
```

```yaml
apiVersion: aadpodidentity.k8s.io/v1
kind: AzureIdentityBinding
metadata:
  name: az-id-2-binding
spec:
  azureIdentity: az-id-2
  selector: az-id-combined
```

### Pod

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: demo
  labels:
    aadpodidbinding: az-id-combined
...
```

## Pods using unauthorized AzureIdentities

By default, aad-pod-identity matches pods with `AzureIdentities` across all namespaces. That means that a malicious pod could assign itself an unauthorized pod identity label and acquire a token with that particular `AzureIdentity`. This scenario can be mitigated by performing the following:

> - Deploy aad-pod-identity with [namespaced mode](README.namespaced.md). This will restrict pods from binding with `AzureIdentities` across different namespaces. Note that `AzureIdentities` and `AzureIdentityBindings` must deploy to the same namespace as your workload when using namespaced mode.

Reviewer: ```suggestion
- Deploy aad-pod-identity with [namespaced mode](README.namespaced.md). This will restrict pods from binding with `AzureIdentities` across different namespaces. Note that `AzureIdentities` and `AzureIdentityBindings` must be deployed in the same namespace as your workload when using namespaced mode.
```
```suggestion
- Deploy aad-pod-identity with [namespaced mode](README.namespaced.md). This will restrict pods from binding with `AzureIdentities` across different namespaces. Note that `AzureIdentities` and `AzureIdentityBindings` must deploy to the same namespace as your workload when using namespaced mode.
```
